tests/test_json_stream.py

```diff
@@ -0,0 +1,495 @@
+
+    assert json.loads(complex_dict)
+
+    bytes_io_chunk = BytesChunkIO(complex_dict.encode(), chunk_size)
+
+    stream = adafruit_json_stream.load(bytes_io_chunk)
+
+    list_1 = stream["list_1"]
+    dict_1 = next(list_1)
+    sub_list = dict_1["sub_list"]
+    assert next(sub_list) == "a"
+
+    assert bytes_io_chunk.get_chunks_read() == expected_chunks
+    assert math.ceil(len(complex_dict) / chunk_size) >= (expected_chunks / 4)
+
+
+def test_incomplete_json_raises():
+    """Test incomplete json raises."""
+
+    data = """
+    {
+        "field_1": 1
+    """
+
+    with pytest.raises(json.JSONDecodeError):
+        json.loads(data)
+
+    stream = adafruit_json_stream.load(BytesChunkIO(data.encode()))
+
+    with pytest.raises(EOFError):
+        stream["field_2"]
+
+
+def test_as_object(complex_dict):
+    """Test loading a complex dict and grabbing parts as objects."""
+
+    assert json.loads(complex_dict)
+
+    stream = adafruit_json_stream.load(BytesChunkIO(complex_dict.encode()))
+
+    list_1 = stream["list_1"]
+    dict_1 = next(list_1)
+    assert dict_1["sub_dict"].as_object() == {
+        "sub_dict_id": 1.1,
+        "sub_dict_name": "one point one",
+    }
+    assert dict_1["sub_list"].as_object() == ["a", "b", "c"]
+    dict_2 = next(list_1)
+    assert dict_2.as_object() == {
+        "dict_id": 2,
+        "dict_name": "two",
+        "sub_dict": {"sub_dict_id": 2.1, "sub_dict_name": "two point one"},
+        "sub_list": ["d", "e", "f"],
+    }
+    assert stream["list_2"].as_object() == [
+        {
+            "dict_id": 3,
+            "dict_name": "three",
+            "sub_dict": {"sub_dict_id": 3.1, "sub_dict_name": "three point one"},
+            "sub_list": ["g", "h", "i"],
+        },
+        {
+            "dict_id": 4,
+            "dict_name": "four",
+            "sub_dict": {"sub_dict_id": 4.1, "sub_dict_name": "four point one"},
+            "sub_list": ["j", "k", "l"],
+        },
+    ]
+
+
+def test_as_object_stream(dict_with_all_types):
+
+    assert json.loads(dict_with_all_types)
+
+    stream = adafruit_json_stream.load(BytesChunkIO(dict_with_all_types.encode()))
+
+    assert stream.as_object() == {
+        "bool": True,
+        "dict": {"key": "value"},
+        "float": 1.1,
+        "int": 1,
+        "list": [1, 2, 3],
+        "null": None,
+        "string": "string",
+    }
+
+
+def test_as_object_that_is_partially_read_raises(complex_dict):
+    """Test loading a complex dict and grabbing partially read raises."""
+
+    assert json.loads(complex_dict)
+
+    stream = adafruit_json_stream.load(BytesChunkIO(complex_dict.encode()))
+
+    list_1 = stream["list_1"]
+    dict_1 = next(list_1)
+    assert dict_1["dict_id"] == 1
+    with pytest.raises(BufferError):
+        dict_1.as_object()
+
+
+def test_as_object_grabbing_multiple_subscriptable_levels_twice_raises(complex_dict):
+    """Test loading a complex dict and grabbing multiple subscriptable levels twice raises."""
```

Comment:
Should this work not fail, assuming the keys are ordered and the user hasn't left [asked for later data] the current parent object?
If not then what is the correct usage, to request the parent object separately first and then subscript that?

Reply:
@tyeth challenge accepted and fixed: 1b47d3ef1ee7f05df4a764f955bc3a1c4ea067dd